src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,1209 @@
+                    return true;
+                case SubscriptionConnectionServerMessage.MessageType.EndOfBatch:
+                    await WriteMessageToClientWorker(receivedMessage, currentWorker);
+                    return false;
+                default:
+                    await HandleServerMessageFromServerWorkers(workers, receivedMessage, currentWorker);
+                    return false;
+            }
+        }
+
+        //TODO: egor handle double code with initasync
+        private async Task HandleServerMessageFromServerWorkers(List<SubscriptionServerWorker> workers, SubscriptionConnectionServerMessage receivedMessage, SubscriptionServerWorker currentWorker)
+        {
+            switch (receivedMessage.Type)
+            {
+                case SubscriptionConnectionServerMessage.MessageType.Confirm:
+                    break;
+                case SubscriptionConnectionServerMessage.MessageType.ConnectionStatus:
+                    switch (receivedMessage.Status)
+                    {
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.None:
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.InUse:
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.Closed:
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.NotFound:
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.ForbiddenReadOnly:
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.Forbidden:
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.Invalid:
+                            //TODO: egor throw on some of those statuses ?
+                            CancellationTokenSource.Cancel();
+                            break;
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.ConcurrencyReconnect:
+                            // drop subscription on other workers
+                            foreach (var worker in workers)
+                            {
+                                if (worker != currentWorker)
+                                {
+                                    // propagate the error so the other server workers fail
+                                    var t = SendClientReplyToServerWorker(worker, new SubscriptionConnectionClientMessage()
+                                    {
+                                        ChangeVector = _lastChangeVector,
+                                        Type = SubscriptionConnectionClientMessage.MessageType.Acknowledge
+                                    });
+                                }
+                            }
+
+                            throw new SubscriptionChangeVectorUpdateConcurrencyException(receivedMessage.Message);
+                        case SubscriptionConnectionServerMessage.ConnectionStatus.Redirect:
```

Comment:
You have a *lot* of code to manage the subscription connection, and a lot of that is duplicated.